core/src/test/java/io/questdb/test/griffin/SqlOptimiserTest.java

```diff
@@ -170,6 +171,312 @@ public void testAliasAppearsInFuncArgs6() throws Exception {
+            ddl(orderByAdviceDdl.replace(" t ", " t1 "));
+            ddl(orderByAdviceDdl.replace(" t ", " t2 "));
+
+            final String query = "SELECT t1.s, t1.ts, t2.s, t2.ts\n" +
+                    "FROM t1\n" +
+                    "ASOF JOIN t2 ON t1.s = t2.s\n" +
+                    "WHERE t1.ts in '2023-09-01T00:00:00.000Z' AND t1.ts <= '2023-09-01T01:00:00.000Z'\n" +
+                    "ORDER BY t2.s, t2.ts\n" +
+                    "LIMIT 1000000;";
+
+            assertQuery("select-choose t1.s s, t1.ts ts, t2.s s1, t2.ts ts1 from (select [s, ts] from t1 timestamp (ts) asof join select [s, ts] from t2 timestamp (ts) on t2.s = t1.s where ts in '2023-09-01T00:00:00.000Z' and ts <= '2023-09-01T01:00:00.000Z') order by s1, ts1 limit 1000000", query);
+            assertPlan(query, "Limit lo: 1000000\n" +
+                    "    Sort\n" +
+                    "      keys: [s1, ts1]\n" +
+                    "        SelectedRecord\n" +
+                    "            AsOf Join Light\n" +
+                    "              condition: t2.s=t1.s\n" +
+                    "                DataFrame\n" +
+                    "                    Row forward scan\n" +
+                    "                    Interval forward scan on: t1\n" +
+                    "                      intervals: [(\"2023-09-01T00:00:00.000000Z\",\"2023-09-01T00:00:00.000000Z\")]\n" +
+                    "                DataFrame\n" +
+                    "                    Row forward scan\n" +
+                    "                    Frame forward scan on: t2\n");
+            try (RecordCursorFactory factory =  select(query, sqlExecutionContext)) {
+            }
+        });
+    }
+
+    @Test
+    public void testOrderByAdviceWorksWithRegularJoin() throws Exception {
+        assertMemoryLeak(() -> {
+            ddl(orderByAdviceDdl.replace(" t ", " t1 "));
+            ddl(orderByAdviceDdl.replace(" t ", " t2 "));
+
+            final String query = "SELECT t1.s, t1.ts, t2.s, t2.ts\n" +
+                    "FROM t1\n" +
+                    "JOIN t2 ON t1.s = t2.s\n" +
+                    "WHERE t1.ts in '2023-09-01T00:00:00.000Z' AND t1.ts <= '2023-09-01T01:00:00.000Z'\n" +
+                    "ORDER BY t1.s, t1.ts\n" +
+                    "LIMIT 1000000;";
+
+            assertQuery("select-choose t1.s s, t1.ts ts, t2.s s1, t2.ts ts1 from (select [s, ts] from t1 timestamp (ts) join select [s, ts] from t2 timestamp (ts) on t2.s = t1.s where ts in '2023-09-01T00:00:00.000Z' and ts <= '2023-09-01T01:00:00.000Z') order by s, ts limit 1000000", query);
+            assertPlan(query, "Limit lo: 1000000\n" +
+                    "    Sort\n" +
+                    "      keys: [s, ts]\n" +
+                    "        SelectedRecord\n" +
+                    "            Hash Join Light\n" +
+                    "              condition: t2.s=t1.s\n" +
+                    "                SortedSymbolIndex\n" +
```

Comment:
What is the benefit of sorting here if the sort will happen outside anyway?

Reply:
In my opinion, it should either not have the top level sort or use `Interval forward scan on` on both tables. Both options make sense but doing both of them is definitely suboptimal.